google/web-stories-wp
# Fix Page Templates usage

## Context



It was reported by us and also by users that the "Saved Templates" function is broken in the last release due to [multiple bugs](https://github.com/google/web-stories-wp/issues/9294#issuecomment-936609703).

## Summary



Since this is actually three bugs in one, this PR contains some JavaScript changes but also some PHP changes.

In essence, it makes saving custom page templates work again.

See below for details.

## Relevant Technical Choices



**Note:** see also individual commits for a better understanding.

**Don't fetch page templates if `nextTemplatesToFetch` is `false`**

This is self-explanatory. Basically, this would lead to REST API requests being made with `?page=false`, which will be rejected by the API because `false` clearly is not an `integer`.

**Ensure `TemplateList` doesn't crash when `pages` is `null`**

This is self-explanatory. `pages` is always `null` at the beginning. And when REST API fetching fails, it will stay `null`. However, the component didn't support this and crashed when `pages` wasn't an array. So this adds some extra hardening.

**Always show `TemplateSave` button**

Again, self-explanatory. Always show the button - even if templates haven't been loaded yet.

**Improve service registration to ensure correct order**

This change allows us to clearly define dependencies between services using `HasRequirements`, without having to use different action priorities at the same time.

**Example:** `Settings`, `Story_Post_Type` and `Page_Template_Post_Type` are all delayed to be registered on `init`, with priority 10.

However, because `Story_Post_Type` depends on `Settings`, it will have priority 11 to ensure it runs after.

And because `Page_Template_Post_Type` depends on `Story_Post_Type`, it will have new priority 12 to ensure the `Story_Post_Type` service is properly registered first.

This, together with extended usage of `HasRequirements` throughout the code base makes dependency resolution more robust.

**Remove unnecessary abstraction for REST controllers**

`Autosaves_Controller` and `Lock_Controller` were both abstract classes, theoretically making it easier to implement them for multiple post types. However, there are no plans for ever using those for any post types other than web-story, so having abstract classes is overkill.

The code has been merged into `Stories_Autosaves_Controller` and `Stories_Lock_Controller`, respectively.

**E2E tests**

To prevent regressions, this PR adds an e2e test verifying that custom page templates can be loaded & saved.

## To-do



N/A

## User-facing changes



* The "Save page as template" button now doesn't have a hover effect when disabled
* The "Saved Templates" panel is no longer broken

## Testing Instructions






This PR can be tested by following these steps:

1. Create a new story
1. Go to page templates tab
2. Verify that the "Save page as template" button is disabled
3. Add an element to the page
4. Verify that the button is no longer disabled and allows you to create a new page template


## Reviews

### Does this PR have a security-related impact?



No

### Does this PR change what data or activity we track or use?



No

### Does this PR have a legal-related impact?



No

## Checklist




---



Fixes #9294
